packages/protocol/contracts/layer1/preconf/impl/PreconfRouter2.sol

```diff
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.24;
+
+import "src/layer1/based/ITaikoInbox.sol";
+import "src/layer1/based/IProposeBatch.sol";
+import "src/layer1/preconf/iface/IPreconfRouter2.sol";
+import "src/shared/common/EssentialContract.sol";
+import "src/layer1/preconf/iface/ILookaheadStore.sol";
+import "src/layer1/preconf/iface/IPreconfWhitelist.sol";
+import "src/layer1/preconf/libs/LibPreconfUtils.sol";
+import "@eth-fabric/urc/IRegistry.sol";
+
+/// @title PreconfRouter2
+/// @custom:security-contact security@taiko.xyz
+contract PreconfRouter2 is IPreconfRouter2, EssentialContract {
+    ILookaheadStore public immutable lookaheadStore;
+    IPreconfWhitelist public immutable preconfWhitelist;
+    IProposeBatch public immutable iProposeBatch;
+    IRegistry public immutable urc;
+    address public immutable preconfSlasher;
+    address public immutable fallbackPreconfer;
+
+    uint256[50] private __gap;
+
+    constructor(
+        address _resolver,
+        address _lookaheadStore,
+        address _preconfWhitelist,
+        address _iProposeBatch,
+        address _preconfSlasher,
+        address _urc,
+        address _fallbackPreconfer
+    )
+        EssentialContract(_resolver)
+    {
+        lookaheadStore = ILookaheadStore(_lookaheadStore);
+        preconfWhitelist = IPreconfWhitelist(_preconfWhitelist);
+        iProposeBatch = IProposeBatch(_iProposeBatch);
+        preconfSlasher = _preconfSlasher;
+        urc = IRegistry(_urc);
+        fallbackPreconfer = _fallbackPreconfer;
+    }
+
+    function init(address _owner) external initializer {
+        __Essential_init(_owner);
+    }
+
+    /// @inheritdoc IProposeBatch
+    function v4ProposeBatch(
+        bytes calldata _params,
+        bytes calldata _txList,
+        bytes calldata _lookaheadData
+    )
+        external
+        nonReentrant
+        returns (ITaikoInbox.BatchInfo memory info_, ITaikoInbox.BatchMetadata memory meta_)
+    {
+        uint256 epochTimestamp = LibPreconfUtils.getEpochTimestamp();
+        bytes26 currentLookaheadHash = lookaheadStore.getLookaheadHash(epochTimestamp);
+
+        // Try fetching the lookahead for the next epoch.
+        // This call fails if the lookahead for next epoch is not posted, thus requiring the first
+        // preconfer to post the next epoch's lookahead before proposing a batch in the current
+        // epoch.
+        lookaheadStore.getLookaheadHash(epochTimestamp + LibPreconfConstants.SECONDS_IN_EPOCH);
+
+        if (_isEmptyLookahead(epochTimestamp, currentLookaheadHash)) {
+            // The current lookahead is empty, so we use a whitelisted preconfer
```

Comment:
if the current lookahead is empty, can we also force the first proposer to post the lookahead?